### PR TITLE
Add Playground to Lite Custom Element

### PR DESCRIPTION
**Pull request by aliabd**
Adds the `playground` attribute to the `<gradio-lite>` custom element. Which creates an interactive side-by-side code and demo playground. It hides the demo until it's loaded (but shows the loading messages on the top right). It will take the full width and height of it's container, and switch to rows from columns at the md breakpoint. 

Example: 

```html
<gradio-lite playground>
import gradio as gr
gr.Interface(
	fn=lambda x: x,
	inputs=gr.Textbox(),
	outputs=gr.Textbox()
).launch()
</gradio-lite>
```

https://github.com/gradio-app/gradio/assets/9021060/c821a5b7-c526-42b6-b511-867a179a4dd8

To test this run: 

```bash
pnpm --filter @gradio/app build:lite
python -m http.server --directory js/lite
```

This will start a server at `http://localhost:8000`. You can then edit `js/lite/index.html` to add a gradio-lite element with the playground attribute.

(I moved these changes from #7568 to make releasing easier. They need to be released first before we can add them to the website)

### CODE REVIEW

**Comment by abidlabs** [comment]
Thanks for the instructions for reviewing. Taking a look now!

**Comment by abidlabs** [comment]
This looks great and works super well @aliabd! I created a simple test page with 2 Gradio Playground elements and tested it and here's my suggestions based on my tests:

**1. Padding and width**

It seems that the `<gradio-lite playground>` tag adds its own padding, which may undesired in some situations. I would not add any padding/margin/etc and let users control it themselves by wrapping in a div and styling it as needed. Also, when I added some padding to the body, I noticed that the width of the playground element exceeded the width of the page, forcing the user to horizontally scroll. We should set the default width of the playground to be 100% of the parent container so that this does not happen. You can see both of the issues in the index.html page that I uploaded to this PR. 

You can also see it in this screenshot: 

image

**2. Code Parsing**

Right now, the playground tag is sensitive to the whitespace and indentation of the codeblock inside of it. This makes it hard to format the code properly to remove the whitespace. I have to do something like:

```
		<gradio-lite playground>import gradio as gr 
gr.ChatInterface(lambda x,y:x).launch()
		</gradio-lite>

```

whereas this should work:

```
		<gradio-lite playground>
                  import gradio as gr 
                  gr.ChatInterface(lambda x,y:x).launch()
		</gradio-lite>
```

It would be great if we could trim any leading or trailing new lines and also remove any indentation/whitespace that can be uniformly removed all of the lines. In other parts of the codebase, we actually do this using the `inspect.cleandoc` function in python but here's well need to use js I assume

**3. Small stylistic feedback**

While loading...

image

* As we noted together, the font for "Code" looks ugly until the Interface has fully loaded
* Instead of the word "Code", what about "app.py" to indicate that its a Python file
* I also think the shade of gray that shows the "loading pyodide" messages should be a bit lighter

image

After its loaded... I think the green dot, the word interactive, and update button are too crowded. I would instead:

* Remove the green dot (the green is too bright and clashes with gradio's theme) and instead put a gray lightning icon and the word interactive in a light gray.
* I would move the "update button" to the right panel and remove preview. I think that's obvious given the context
* Instead of "Update ->", I would reword as "Run" followed by the play button

**4. CPU and memory usage**

I noticed that while pyodide is loading, it consumes about 20% of my CPU and after the page is loaded, it consumes about 3-4 gb of ram. The CPU usage is not a problem, the RAM we should keep an eye out. I also opened a regular Gradio Hugging Face Space and it consumed about the same, so this might be inevitable. But just noting here.

File: js/app/src/lite/index.ts [hunk 1]
```diff
@@ -116,55 +118,104 @@ export function create(options: Options): GradioAppController {
...
+			app.$on("code", (code) => {
+				options.code = code.detail.code;
+				worker_proxy
+					.runPythonCode(options.code)
+					.then(launchNewApp(true))
+					.catch((e) => {
+						showError(e);
+						throw e;
+					});
+			});
```
Comment (whitphx) [inline js/app/src/lite/index.ts:148]: What if we move this snippet to the `<Playground />` component and call `worker_proxy.runPythonCode(...)` in `<Playground />` directly?
The logic around code updates will be encapsulated into `<Playground />` and we won't need to dispatch and catch the `code` event across the components.

**Comment by whitphx** [comment]
@abidlabs 
> 4. CPU and memory usage

If you are testing with multiple `<gradio-lite>` tags, can you try the SharedWorker mode by adding the `shared-worker` attribute to **all** the `<gradio-lite>` as below?
```html
		<gradio-lite playground shared-worker>
import gradio as gr
gr.Interface(fn=lambda x: x, inputs=gr.Textbox(), outputs=gr.Textbox()).launch()
		</gradio-lite>

		<gradio-lite playground shared-worker>
import gradio as gr
gr.ChatInterface(lambda x,y:x).launch()
		</gradio-lite>
```
I tried it with the `js/lite/index.html` on this branch and it reduced the RAM usage from 1.2GB to 650MB.

The SharedWorker mode can improve the multi-worker situation, while the high RAM usage per a single worker is still a big problem though. 


(BTW the RAM usage in my env (M1 Mac, Chrome) with 2 workers was 1.2GB. It's still high, but the number may depend on the env somehow?)

**Comment by whitphx** [comment]
Related to the styling issue @abidlabs addressed above:
The layout in a narrow window should be considered?
CleanShot 2024-03-12 at 12 48 31@2x

---

* Keep the 2 columns side-by-side? [Shinylive](https://shinylive.io/py/examples/) does this as below even for mobile display width. Codesandbox and JsFiddle also do the same.
    
* Stack the 2 panes vertically, as it currently does?
  * If so, the header padding in the preview pane should be removed when stacked.

File: js/theme/src/global.css [hunk 2]
```diff
@@ -1,9 +1,9 @@
 /*
  Custom element styles
  */
-gradio-lite {
+/* gradio-lite {
```
[the remark this answers is not in the document]
Comment (whitphx) [reply js/theme/src/global.css:4]: I remember I added this because it was necessary for the embedded app to be styled correctly atm.
Probably it's behaving the same as the following style which is for the normal Gradio.
https://github.com/gradio-app/gradio/blob/aba44707af20b78a7b87f5f7bd3f189d971afc71/js/app/index.html#L95

If Lite works well without it now, this can be removed.

**Comment by whitphx** [comment]
> 2. Code Parsing
> we actually do this using the `inspect.cleandoc` function in python but in this case, we'll need to use js I assume

Or [`textwrap.dedent`](https://docs.python.org/3/library/textwrap.html#textwrap.dedent), while I'm not sure about the diff actually.
For JS/TS, these libs are available.
* [dedent](https://www.npmjs.com/package/dedent)
* [dedent-ts](https://www.npmjs.com/package/ts-dedent)

File: js/app/src/lite/Playground.svelte [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	// $: dispatch("code", { code });
+
+	let dummy_elem: any = { classList: { contains: () => false } };
+	let dummy_gradio: any = { dispatch: (_) => {} };
+
+	let loading_text = "";
+	export let loaded = false;
+	worker_proxy?.addEventListener("progress-update", (event) => {
+		loading_text = (event as CustomEvent).detail + "...";
+		if (loading_text === "Initialization completed...") {
```
[the remark this answers is not in the document]
Comment (aliabd) [reply js/app/src/lite/Playground.svelte]: Done, good call.

File: js/app/src/lite/Playground.svelte [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+				<Code
+					bind:value={code}
+					label=""
+					language="python"
+					target={dummy_elem}
+					gradio={dummy_gradio}
+					lines={10}
+					interactive={false}
+				/>
+			{/if}
```
[the remark this answers is not in the document]
Comment (aliabd) [reply js/app/src/lite/Playground.svelte]: This doesn't work actually, and I'm not sure why. But it doesn't make the code component interactive after loaded=true.

File: js/app/src/lite/Playground.svelte [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+	export let host: string | null;
+	export let src: string | null;
+
+	export let code: string | undefined;
+	export let error_display: SvelteComponent | null;
+
+	const dispatch = createEventDispatcher();
+	// $: dispatch("code", { code });
+
+	let dummy_elem: any = { classList: { contains: () => false } };
```
[the remark this answers is not in the document]
Comment (aliabd) [reply js/app/src/lite/Playground.svelte:39]: yeah I wrote these dummy_elem and dummy_gradio hacks for the playground tab in the website, so I could use the code component outside the context of gradio. And just copied the lines here. 

@pngwn is there a better way to do this?

File: js/theme/src/global.css (hunk 2 above)
[the remark this answers is not in the document]
Comment (aliabd) [reply js/theme/src/global.css]: Restored it

File: js/app/src/lite/index.ts (hunk 1 above)
Comment (aliabd) [reply js/app/src/lite/index.ts]: That's what I intended to do initially but couldn't figure out how. I can runPythonCode inside <Playground /> but I can't launchNewApp afterwards (since the whole playground is the app). running the code just kills the port. How do I work around this?

**Comment by aliabd** [comment]
@abidlabs i implemented all the changes, please take another look. Didn't do anything about 4 but will keep an eye on it. Though, like @whitphx said sharedworker mode does help a lot when there are multiple apps running.

**Comment by abidlabs** [comment]
Thanks @aliabd! Will make another pass today

**Review by abidlabs — APPROVED** [review]
Really nice stuff @aliabd! This looks really good. 

Only other piece of feedback I have is that it would be really nice if there was a keyboard shortcut that re-ran the code instead of having to go back and forth from typing and clicking on the Run button. I'd suggest ctrl/cmd + enter (same as jupyter notebooks)

Otherwise LGTM.

**Comment by abidlabs** [comment]
Nit: the font for "app.py" should be a monospace font imo because its a filename

image

**Comment by abidlabs** [comment]
Also perhaps there should be a better indication when the Python file is broken. I like that the old demo is displayed instead of an ugly error. But we should also indicate that that there is an error in some way. Like maybe a gray ⚠️ appears before the filename and hovering over it tells you the error. 

https://github.com/gradio-app/gradio/assets/1778297/dce766c6-75d9-4729-ba36-d44a3a5b9e0f

(This can also be in a separate PR if needed)

File: js/app/src/lite/index.ts (hunk 1 above)
Comment (whitphx) [reply js/app/src/lite/index.ts]: Ah sorry you are right, I forgot about `launchNewApp`.
Maybe we can fix `lite/index.ts` and `Playground.svelte` so that `<Playground />` is kept mounted when the code is changed and only `<Index />` is refreshed inside it, but the current code should be good for now.

File: js/lite/index.html [hunk 6]
```diff
@@ -14,36 +14,81 @@
 			href="https://fonts.gstatic.com"
 			crossorigin="anonymous"
 		/>
+		<script src="https://cdn.tailwindcss.com"></script>
```
[the remark this answers is not in the document]
Comment (aliabd) [reply js/lite/index.html]: fixed now. removed tailwind and used plain css instead. we don't use tailwind on actual gradio apps

**Comment by aliabd** [comment]
@abidlabs thanks, i added the shortcut and fixed app.py font. try it out! 

So we currently do show errors instead of the interface. The reason it's not showing an error for you here though is because you didn't launch the demo. so the code is running fine but there's no new app to replace the old one. Thanks to @whitphx in https://github.com/gradio-app/gradio/pull/7572 we do show a console error in this case. 

Anyway i do definitely think there's a better way to show errors and really like your suggestion. will work on it on a separate PR. 

This is good to merge now. The remaining questions are all around refactoring and we can deal with them separately after.

File: js/app/src/lite/Playground.svelte (hunk 5 above)
[the remark this answers is not in the document]
Comment (pngwn) [reply js/app/src/lite/Playground.svelte:40]: I think we can just remove this `target` prop  and pass the theme down directly. It isn't used for anything else.

File: js/app/src/lite/Playground.svelte [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+	import type { ThemeMode } from "../types";
+	import { mount_css as default_mount_css } from "../css";
+	import type { api_factory } from "@gradio/client";
+	import type { WorkerProxy } from "@gradio/wasm";
+	import { SvelteComponent, createEventDispatcher } from "svelte";
+	import Code from "@gradio/code";
+	import ErrorDisplay from "./ErrorDisplay.svelte";
+	import lightning from "../images/lightning.svg";
+	import play from "../images/play.svg";
+	import { n } from "@storybook/theming/dist/create-3ae9aa71";
```
Comment (pngwn) [inline js/app/src/lite/Playground.svelte:12]: I think is included by mistake (vscode auto-import probably did this.)

File: js/app/src/lite/index.ts [hunk 8]
```diff
@@ -87,6 +89,26 @@ export function create(options: Options): GradioAppController {
...
+				minIndent =
+					minIndent !== null ? Math.min(minIndent, indentLength) : indentLength;
+			}
+		});
+		if (minIndent === null || minIndent === 0) {
+			return code.trim();
+		}
+		const normalizedLines = lines.map((line) => line.substring(minIndent));
+		return normalizedLines.join("\n").trim();
+	}
```
Comment (pngwn) [inline js/app/src/lite/index.ts:108]: Can we make this snake_case?

**Review by pngwn — APPROVED** [review]
Few small things but looks great!

File: js/app/src/lite/Playground.svelte (hunk 4 above)
[the remark this answers is not in the document]
Comment (pngwn) [reply js/app/src/lite/Playground.svelte]: The reason this is a problem is because `interactive` isn't interactive (oh the irony).

We set the 'interactivity' by changing the `readonly` status of the codemirror instance, which is controlled via an extension (just codemirror things, everything is an extension).

The fix for this is weird but we would need to do something like this (untested but i think should work with small tweaks):

```ts
const _readonly = new Compartment();
const _editable = new Compartment();

// when we set the initial extension state (around line 200)

const extensions: Extension[] = [
  _readonly.of(EditorState.readOnly.of(readonly)),
  _editable.of(EditorView.editable.of(!readonly)),
  EditorView.contentAttributes.of({ "aria-label": "Code input container" })
];


// when the readonly prop changes

function set_readonly(view: EditorView, readonly: boolean): void {
  view.dispatch({
    effects: [
        _readonly.reconfigure(EditorState.readOnly.of(readonly)),
        _editable.reconfigure(EditorView.editable.of(!readonly))
    ]
  });
}

$: set_readonly(view, readonly);
```

If you want you could merge this as is and we can fix this in a followup, not a huge deal.

File: js/app/src/lite/Playground.svelte (hunk 5 above)
[the remark this answers is not in the document]
Comment (pngwn) [reply js/app/src/lite/Playground.svelte:39]: Actually just changed it to this which is already passed down by gradio. 

```
type ThemeMode = "system" | "light" | "dark";

export let theme_mode: ThemeMode;

let dark_mode = theme_mode === "dark";
```

I think the code component was created just before we started passing this down.

File: js/app/src/lite/Playground.svelte [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+	export let src: string | null;
+
+	export let code: string | undefined;
+	export let error_display: SvelteComponent | null;
+
+	const dispatch = createEventDispatcher();
+	// $: dispatch("code", { code });
+
+	let dummy_elem: any = { classList: { contains: () => false } };
+	let dummy_gradio: any = { dispatch: (_) => {} };
```
[the remark this answers is not in the document]
Comment (pngwn) [reply js/app/src/lite/Playground.svelte:40]: We just need to make sure it is handled appropriately in the code component.
